File: wasm/wcsjs/src/wrapper.js
addOnPostRun(() => {
  function WCS() {
    this._getWCS = Module.cwrap("getWcs", "number", ["String", "number"]);

    this._hasCelestial = Module.cwrap("hasCelestial", "number", [
      "number"
    ]);

    this._pix2sky = Module.cwrap("pix2sky", "number", [
      "number",
      "number",
      "number",
      "number"
    ]);

    this._sky2pix = Module.cwrap("sky2pix", "number", [
      "number",
      "number",
      "number",
      "number"
    ]);
  }

  function string2buffer(str) {
    let buffer = new ArrayBuffer(str.length);
    let view = new Uint8Array(buffer);
    for (let i = 0; i < str.length; i += 1) {
      view[i] = str.charCodeAt(i);
    }

    return buffer;
  }

  function splice(str1, index, remove, str2) {
    return str1.slice(0, index) + str2 + str1.slice(index + Math.abs(remove));
  }

  function toHeader(wcsObj) {
    let header = [];
    let line =
      "                                                                                ";

    for (let card in wcsObj) {
      let value = wcsObj[card];
      if (value === undefined || value === null) {
        continue;
      }

      if (typeof value === "string" && value !== "T" && value !== "F") {
        value = "'" + value + "'";
      }

      let entry = splice(line, 0, card.length, card);
      entry = splice(entry, 8, 1, "=");
      entry = splice(entry, 10, value.toString().length, value);

      header.push(entry);
    }

    return header.join("\n");
  }

  const wcsRegEx = [
    "DATE-OBS",
    "EQUINOX",
    "WCSAXES",
    "RADESYS",
    "LONPOLE",
    "LATPOLE",
    /NAXIS\d*/,
    /CTYPE\d+/,
    /CRPIX\d+/,
    /CRVAL\d+/,
    /CUNIT\d+/,
    /CDELT\d+/,
    /CD.+/,
    /PC.+/,
    /PV.+/,
    /CROTA\d+/
  ];

  // Defines the maximum number of coordinate calculations to perform per WCSlib call. A larger number will result in a faster conversion, but will require more memory
  // Due to the intermediate arrays required, each coordinate calculation requires 68 bytes of memory to be allocated. A default limit of 5000 coordinates per iteration
  // requires 340k of memory.
  const MAX_COORDS = 5000;

  // Initialize a WCS object using either a string or object of key value pairs
  WCS.prototype.init = function(header) {
    var headerStr, nkeyrec, nHeaderBytes, headerPtr, headerHeap;

    if (typeof header === "object") {
      header = toHeader(header);
    }

    // Split the string into an array and filter based on the WCS regular expressions
    var headerArray = header.match(/.{1,80}/g);
    headerArray = headerArray.filter(function(line) {
      // Extract the keyword
      var keyword = line.slice(0, 8).trim();

      for (var i = 0; i < wcsRegEx.length; i += 1) {
        var regEx = wcsRegEx[i];
        if (keyword.match(regEx)) {
          return true;
        }
      }

      return false;
    });
    headerStr = headerArray.join("\n");
    console.log(headerStr);

    nkeyrec = headerArray.length;
    header = string2buffer(headerStr);

    // Allocate string on Emscripten heap and get byte offset
    nHeaderBytes = header.byteLength;
    headerPtr = Module._malloc(nHeaderBytes);
    headerHeap = new Uint8Array(Module.HEAPU8.buffer, headerPtr, nHeaderBytes);
    headerHeap.set(new Uint8Array(header));

    // Allocate memory on the Emscripten heap for coordinates
    this.coordinatePtr = Module._malloc(16);

    // Use byte offset to pass header string to libwcs
    this.wcsPtr = this._getWCS(headerHeap.byteOffset, nkeyrec);
    this.isValid = this.wcsPtr ? true : false;
    // console.log(this.isValid, this.wcsPtr);
  };

  WCS.prototype.pix2sky = function(x, y) {
    const retVal = this._pix2sky(this.wcsPtr, x, y, this.coordinatePtr);
    const world = new Float64Array(
      Module.HEAPF64.buffer,
      this.coordinatePtr,
      2
    );
    return new Float64Array(world);
  };

  WCS.prototype.hasCelestial = function() {
    return this._hasCelestial(this.wcsPtr);
  }

  WCS.prototype.sky2pix = function(ra, dec) {
    this._sky2pix(this.wcsPtr, ra, dec, this.coordinatePtr);
    const pixcrd = new Float64Array(
      Module.HEAPU8.buffer,
      this.coordinatePtr,
      2
    );
    return pixcrd.slice(0);
  };

  if (typeof module !== "undefined" && module.exports) {
    module.exports = WCS;
  } else {
    window.WCS = WCS;
  }
  console.log("WCS ready");
});

File: wasm/wcsjs/src/wrapper.c
#include <stdio.h>
#include <wcslib/wcshdr.h>
#include <wcslib/wcs.h>

struct wcsprm *getWcs(char *header, int nkeyrec)
{
	int relax = WCSHDR_all, ctrl = 0;
	int nreject, nwcs;
	struct wcsprm *wcs;

	int retCode = wcspih(header, nkeyrec, relax, ctrl, &nreject, &nwcs, &wcs);
	// printf("HERE LNG LAT %d %d", wcs->lng, wcs->lat);
	// printf("HERE2: %d, %d, %d, %d, %f, %f, %f, %f, %f, %f, %f, %f, %f, %f, %f, %f, %f, %f", nreject, nwcs, wcs->naxis, wcs->altlin, wcs->crpix[0], wcs->crpix[1], wcs->crval[0], wcs->crval[1], wcs->pv[0], wcs->pv[1], wcs->pv[2], wcs->pv[3], wcs->pc[0], wcs->pc[1], wcs->pc[2], wcs->pc[3], wcs->cdelt[0], wcs->cdelt[1]);

	if (retCode || nwcs < 1 || wcs->naxis != 2)
		return NULL;

	return wcs;
}

int hasCelestial(struct wcsprm *wcs) {
	// printf("LNG LAT %d %d %d %d", wcs->lng, wcs->lat, wcs->types[0], wcs->types[1]);
	return wcs->lng >= 0 && wcs->lat >= 0 ? 1 : 0;
}

int pix2sky(struct wcsprm *wcs, double x, double y, double *world)
{
	double imgcrd[2], phi[2], theta[2];
	int status[1];
	double pixcrd[2] = {x, y};

	wcsp2s(wcs, 1, 2, pixcrd, imgcrd, phi, theta, world, status);
	return status[0];
}

int sky2pix(struct wcsprm *wcs, double ra, double dec, double *pixcrd)
{
	double imgcrd[2], phi[2], theta[2];
	int status[1];
	double world[2] = {ra, dec};

	wcss2p(wcs, 1, 2, world, phi, theta, imgcrd, pixcrd, status);
	return status[0];
}
